### ISSUE TITLE
[bug] Asset paths in CSS break if publicPath isn’t defined in bud.config.mjs when upgrading to Bud 6.1

### ISSUE THREAD

**Issue by tedw** [roots/bud#1547]
### Agreement

- [X] This is not a duplicate of an existing issue
- [X] I have read the [guidelines for Contributing to Roots Projects](https://github.com/roots/.github/blob/master/CONTRIBUTING.md)
- [X] This is not a personal support request that should be posted on the [Roots Discourse](https://discourse.roots.io/) community

### Describe the issue

After a lot of trial and error, I realized I needed to add `.setPublicPath('../')` to `bud.config.mjs` to get things working in Bud 6.1 and above. I noticed this was added to the example config two weeks ago (https://github.com/roots/sage/commit/16d99623a90eb4d48fc97bbc8b2121f64c12d9f9) but didn’t see any mention in the docs or release notes. Would be nice if it could be added to the [Bud 6.1 release notes](https://bud.js.org/blog/6.1.0), or wherever you all think makes the most sense.

FWIW it appears the syntax here is breaking the public path in production when `.setPublicPath()` isn’t set:
https://github.com/roots/bud/blob/200bad5add181b2447c9eedff9677d354d37d84a/sources/%40roots/sage/src/acorn/extension.ts#L49-L51

I’m not exactly sure why but adding curly braces seems to fix it  🤷‍♂️ (I edited `node_modules/@roots/sage/lib/acorn/extension.js` locally to test):

```js
this.app.hooks.action('config.after', async () => {
    this.app.hooks.on('build.output.publicPath', publicPath => {
      this.app.isDevelopment ? `/` : publicPath
    });
});
```

### Expected Behavior

I expected some mention about having to define `.setPublicPath('../')` in the release notes, or for it to work without having to define the public path.

### Actual Behavior

`.setPublicPath('../')` must be added to the Bud config

### Steps To Reproduce

1. Create new Sage 10 theme
1. Add fonts or images in CSS using relative paths
   - I also tried using `~@fonts/` but it made no difference
1. Run `yarn build`
1. View asset paths in generated CSS

Here’s the CSS I used to test:

```
@font-face {
  font-family: "Inter";
  font-style: normal;
  font-weight: 400;
  src: url("../fonts/inter/inter-400.woff2") format("woff2"),
       url("../fonts/inter/inter-400.woff") format("woff");
}

body {
  background-image: url('../images/test.jpg');
}
```

Font files were located in `my-theme/resources/fonts/inter/` and the image was in `my-theme/resources/images/`. 

In Bud 6.0, the generated CSS in `my-theme/public/app.[hash].css` looked like this:

```
src: url(fonts/inter-400.[hash].woff2) format("woff2"),
     url(fonts/inter-400.[hash].woff) format("woff");
```

But after upgrading to Bud 6.1, the generated CSS in `my-theme/public/app/entry.[hash].css` contains root-relative paths:

```
src: url(/fonts/inter-400.[hash].woff2) format("woff2"),
     url(/fonts/inter-400.[hash].woff) format("woff");
```

I would expect them to be relative to the CSS file like this:

```
src: url(../fonts/inter-400.[hash].woff2) format("woff2"),
     url(../fonts/inter-400.[hash].woff) format("woff");
```

### version

6.1.0

### What package manager are you using?

yarn classic

### version

1.22.17

### Configuration

```zsh
I tested using the default config (https://github.com/roots/sage/blob/main/bud.config.js) but converted to `.mjs` as outlined here https://bud.js.org/blog/6.0.0#update-your-config-file-to-use-esm-syntax-and-use-the-extension-mjs
```

### Relevant .budfiles

Not sure if you need this but here’s my `webpack.config.js`

```
module.exports = {
  "entry": {
    "app": {
      "import": [
        "@scripts/app",
        "@styles/app"
      ]
    },
    "editor": {
      "import": [
        "@scripts/editor",
        "@styles/editor"
      ]
    }
  },
  "bail": true,
  "cache": {
    "name": "bud.production",
    "type": "filesystem",
    "version": "renn_micxx62izguk8nwkg0mb_u_",
    "cacheDirectory": "/Users/twhitehead/Sites/example/my-theme/.budfiles/cache/webpack",
    "managedPaths": [
      "/Users/twhitehead/Sites/example/my-theme/node_modules"
    ],
    "buildDependencies": {
      "bud": [
        "/Users/twhitehead/Sites/example/my-theme/package.json",
        "/Users/twhitehead/Sites/example/my-theme/.editorconfig",
        "/Users/twhitehead/Sites/example/my-theme/bud.config.js",
        "/Users/twhitehead/Sites/example/my-theme/composer.json",
        "/Users/twhitehead/Sites/example/my-theme/jsconfig.json",
        "/Users/twhitehead/Sites/example/my-theme/tailwind.config.js",
        "/Users/twhitehead/Sites/example/my-theme/theme.json"
      ]
    }
  },
  "context": "/Users/twhitehead/Sites/example/my-theme",
  "experiments": {
    "buildHttp": {
      "allowedUris": [
        null,
        "https://gist.githubusercontent.com/",
        "https://raw.githubusercontent.com/",
        "https://unpkg.com/",
        "https://cdn.skypack.dev/"
      ],
      "cacheLocation": "/Users/twhitehead/Sites/example/my-theme/.budfiles/bud/modules",
      "frozen": false,
      "lockfileLocation": "/Users/twhitehead/Sites/example/my-theme/.budfiles/bud/bud.lock",
      "upgrade": true
    }
  },
  "externalsType": "var",
  "infrastructureLogging": {
    "console": {
      "Console": {}
    },
    "level": "none"
  },
  "mode": "production",
  "module": {
    "rules": [
      {
        "test": {},
        "exclude": [
          {}
        ],
        "parser": {
          "requireEnsure": false
        }
      },
      {
        "oneOf": [
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/mini-css-extract-plugin/dist/loader.js"
              },
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/css-loader/dist/cjs.js",
                "options": {
                  "importLoaders": 1,
                  "sourceMap": false
                }
              },
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-loader/dist/cjs.js",
                "options": {
                  "postcssOptions": {
                    "syntax": "postcss-scss",
                    "plugins": [
                      [
                        "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-import/index.js"
                      ],
                      [
                        "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-nested/index.js"
                      ],
                      [
                        "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-preset-env/dist/index.cjs",
                        {
                          "stage": 1,
                          "features": {
                            "focus-within-pseudo-class": false
                          }
                        }
                      ]
                    ]
                  },
                  "sourceMap": true
                }
              }
            ]
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/mini-css-extract-plugin/dist/loader.js"
              },
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/css-loader/dist/cjs.js",
                "options": {
                  "importLoaders": 1,
                  "localIdentName": "[name]__[local]___[hash:base64:5]",
                  "modules": true,
                  "sourceMap": false
                }
              }
            ]
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme"
            ],
            "exclude": [
              "/Users/twhitehead/Sites/example/my-theme/node_modules"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/csv-loader/index.js"
              }
            ]
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "type": "asset",
            "generator": {
              "filename": "fonts/[name].[contenthash:6][ext]"
            }
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme"
            ],
            "exclude": [
              "/Users/twhitehead/Sites/example/my-theme/node_modules"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/html-loader/dist/cjs.js"
              }
            ]
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "type": "asset/resource",
            "generator": {
              "filename": "images/[name].[contenthash:6][ext]"
            }
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/babel-loader/lib/index.js",
                "options": {
                  "cacheDirectory": "/Users/twhitehead/Sites/example/my-theme/.budfiles/cache/babel",
                  "presets": [
                    [
                      "/Users/twhitehead/Sites/example/my-theme/node_modules/@babel/preset-env/lib/index.js"
                    ],
                    [
                      "/Users/twhitehead/Sites/example/my-theme/node_modules/@babel/preset-react/lib/index.js"
                    ]
                  ],
                  "plugins": [
                    [
                      "/Users/twhitehead/Sites/example/my-theme/node_modules/@babel/plugin-transform-runtime/lib/index.js",
                      {
                        "helpers": false
                      }
                    ],
                    [
                      "/Users/twhitehead/Sites/example/my-theme/node_modules/@babel/plugin-proposal-object-rest-spread/lib/index.js"
                    ],
                    [
                      "/Users/twhitehead/Sites/example/my-theme/node_modules/@babel/plugin-syntax-dynamic-import/lib/index.js"
                    ],
                    [
                      "/Users/twhitehead/Sites/example/my-theme/node_modules/@babel/plugin-proposal-class-properties/lib/index.js"
                    ]
                  ],
                  "env": {
                    "development": {
                      "compact": false
                    }
                  },
                  "root": "/Users/twhitehead/Sites/example/my-theme"
                }
              }
            ]
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme"
            ],
            "exclude": [
              "/Users/twhitehead/Sites/example/my-theme/node_modules"
            ],
            "type": "json",
            "parser": {}
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "type": "asset/resource",
            "generator": {
              "filename": "images/[name].[contenthash:6][ext]"
            }
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme"
            ],
            "exclude": [
              "/Users/twhitehead/Sites/example/my-theme/node_modules"
            ],
            "type": "json",
            "parser": {}
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "type": "asset/resource",
            "generator": {
              "filename": "images/[name].[contenthash:6][ext]"
            }
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme"
            ],
            "exclude": [
              "/Users/twhitehead/Sites/example/my-theme/node_modules"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/xml-loader/index.js"
              }
            ]
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme"
            ],
            "exclude": [
              "/Users/twhitehead/Sites/example/my-theme/node_modules"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/yml-loader/index.js"
              }
            ]
          },
          {
            "test": {},
            "include": [
              "/Users/twhitehead/Sites/example/my-theme/resources"
            ],
            "use": [
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/mini-css-extract-plugin/dist/loader.js"
              },
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/css-loader/dist/cjs.js",
                "options": {
                  "importLoaders": 1,
                  "sourceMap": false
                }
              },
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-loader/dist/cjs.js",
                "options": {
                  "postcssOptions": {
                    "syntax": "postcss-scss",
                    "plugins": [
                      [
                        "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-import/index.js"
                      ],
                      [
                        "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-nested/index.js"
                      ],
                      [
                        "/Users/twhitehead/Sites/example/my-theme/node_modules/postcss-preset-env/dist/index.cjs",
                        {
                          "stage": 1,
                          "features": {
                            "focus-within-pseudo-class": false
                          }
                        }
                      ]
                    ]
                  },
                  "sourceMap": true
                }
              },
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/resolve-url-loader/index.js",
                "options": {
                  "sourceMap": true
                }
              },
              {
                "loader": "/Users/twhitehead/Sites/example/my-theme/node_modules/sass-loader/dist/cjs.js",
                "options": {
                  "implementation": {
                    "sassNull": {},
                    "sassTrue": {
                      "value": true
                    },
                    "sassFalse": {
                      "value": false
                    },
                    "Logger": {
                      "silent": {}
                    },
                    "info": "dart-sass\t1.53.0\t(Sass Compiler)\t[Dart]\ndart2js\t2.17.3\t(Dart Compiler)\t[Dart]",
                    "types": {},
                    "NULL": {},
                    "TRUE": {
                      "value": true
                    },
                    "FALSE": {
                      "value": false
                    }
                  },
                  "sourceMap": true
                }
              }
            ]
          }
        ]
      }
    ],
    "unsafeCache": false
  },
  "name": "bud",
  "node": false,
  "output": {
    "assetModuleFilename": "[name].[contenthash:6][ext]",
    "chunkFilename": "[name].[contenthash:6].js",
    "filename": "[name].[contenthash:6].js",
    "path": "/Users/twhitehead/Sites/example/my-theme/public",
    "pathinfo": false,
    "publicPath": ""
  },
  "optimization": {
    "emitOnErrors": false,
    "minimize": true,
    "minimizer": [
      "...",
      {
        "options": {
          "test": {},
          "parallel": true,
          "minimizer": {
            "options": {
              "preset": [
                "default",
                {
                  "discardComments": {
                    "removeAll": true
                  }
                }
              ]
            }
          }
        }
      }
    ],
    "runtimeChunk": "single",
    "splitChunks": {
      "cacheGroups": {
        "bud": {
          "chunks": "all",
          "priority": -10,
          "reuseExistingChunk": true,
          "test": {}
        },
        "vendor": {
          "chunks": "all",
          "priority": -20,
          "reuseExistingChunk": true,
          "test": {}
        }
      }
    }
  },
  "parallelism": 150,
  "performance": {
    "hints": false
  },
  "recordsPath": "/Users/twhitehead/Sites/example/my-theme/.budfiles/bud/modules.json",
  "stats": {
    "preset": "errors-warnings"
  },
  "target": "browserslist:/Users/twhitehead/Sites/example/my-theme/package.json",
  "plugins": [
    {
      "patterns": [
        {
          "from": "images",
          "to": "[name].[contenthash:6][ext]",
          "context": "/Users/twhitehead/Sites/example/my-theme/resources",
          "noErrorOnMissing": true
        }
      ],
      "options": {}
    },
    {
      "options": {
        "assetHookStage": null,
        "basePath": "",
        "fileName": "manifest.json",
        "filter": null,
        "map": null,
        "publicPath": "",
        "removeKeyHash": {},
        "sort": null,
        "transformExtensions": {},
        "useEntryKeys": false,
        "useLegacyEmit": false,
        "writeToFileEmit": true
      }
    },
    {
      "_sortedModulesCache": {},
      "options": {
        "filename": "[name].[contenthash:6].css",
        "ignoreOrder": false,
        "runtime": true,
        "chunkFilename": "[name].[contenthash:6].css"
      },
      "runtimeOptions": {
        "linkType": "text/css"
      }
    },
    {
      "options": {
        "enabled": true,
        "verbose": false,
        "extensions": {},
        "ignore": [],
        "remove": {}
      },
      "enabled": true,
      "verbose": false
    },
    {
      "options": {
        "emitHtml": false,
        "publicPath": ""
      },
      "plugin": {
        "name": "EntrypointsManifestPlugin",
        "stage": null
      },
      "name": "entrypoints.json"
    },
    {
      "name": "WordPressExternalsWebpackPlugin",
      "stage": null,
      "externals": {
        "type": "window"
      }
    },
    {
      "plugin": {
        "name": "WordPressDependenciesWebpackPlugin",
        "stage": null
      },
      "manifest": {},
      "usedDependencies": {},
      "fileName": "wordpress.json"
    },
    {
      "plugin": {
        "name": "MergedManifestPlugin"
      },
      "file": "entrypoints.json",
      "entrypointsName": "entrypoints.json",
      "wordpressName": "wordpress.json"
    },
    {
      "resourceRegExp": {}
    },
    {
      "resourceRegExp": {}
    },
    {
      "resourceRegExp": {}
    },
    {
      "resourceRegExp": {}
    }
  ],
  "resolve": {
    "alias": {
      "@src": "/Users/twhitehead/Sites/example/my-theme/resources",
      "@dist": "/Users/twhitehead/Sites/example/my-theme/public",
      "@fonts": "/Users/twhitehead/Sites/example/my-theme/resources/fonts",
      "@images": "/Users/twhitehead/Sites/example/my-theme/resources/images",
      "@scripts": "/Users/twhitehead/Sites/example/my-theme/resources/scripts",
      "@styles": "/Users/twhitehead/Sites/example/my-theme/resources/styles"
    },
    "extensions": [
      ".mjs",
      ".cjs",
      ".js",
      ".jsx",
      ".css",
      ".json",
      ".wasm",
      ".yml",
      ".toml",
      ".scss",
      ".sass"
    ],
    "modules": [
      "/Users/twhitehead/Sites/example/my-theme/resources",
      "node_modules"
    ]
  }
}
```

**Comment by kellymears** [comment]
This can be added to the 6.1 docs, sure. 

> I’m not exactly sure why but adding curly braces seems to fix it 🤷‍♂️

It's because now that function doesn't return anything, which is kind of the same as setting it to `undefined`. I think `app.setPublicPath(undefined)` would do the same thing.

**Comment by tedw** [comment]
Gotcha, that makes sense. For a second I thought I didn't understand how arrow functions worked 😄 Thanks!

**Comment by kellymears** [comment]
Added a note in [`6e0dcd3` (#1549)](https://github.com/roots/bud/pull/1549/commits/6e0dcd36992a028043dee4c464b943d41c62f2b2). Ideally, this behavior is fixed in 6.3.0.

We'll get more specific & explicit about this in the documentation but I can say that in the future bud will expect users not serving assets from web root to define a public path. Right now we characterize it as "if you want to use dynamic imports", but it will be simplifying to just assume it is set. 

I think it is a reasonable requirement and certainly a common one. Some notes I've taken on how similar frameworks handle projects which are housed in a sub directory: 

- [vue cli requires publicPath outside of web root](https://cli.vuejs.org/config/#publicpath)
- [next.js requires publicPath outside of web root](https://nextjs.org/docs/api-reference/next.config.js/basepath)
- [vite requires (?) publicPath outside of web root](https://vitejs.dev/guide/build.html#public-base-path) — Not sure that it is actually required here, but it looks like in cases where it isn't required there is usage of a statically analyzed global (import.meta.env.BASE_URL) which we already support in a few different ways.
- [gatsby requires publicPath outside of web root](https://www.gatsbyjs.com/docs/how-to/previews-deploys-hosting/path-prefix/)
